### DISCUSSION TITLE
InputMethodEngine::keyEvent non-blocking?

### DISCUSSION THREAD

**Discussion by qingxiang-jia** [fcitx/fcitx5#655] [Q&A] [answered]
I just realized that the `keyEvent` function is non-blocking. So if the user types too fast, it's possible the first call to `keyEvent` isn't finish, and the second call already happened. Does this means `keyEvent` and any functions it calls needs to be thread safe?

Since `keyEvent` is non-blocking, does that mean if some function it calls will take a long time (e.g. making a network request), we can just let it block since when the user hits the key again, whether the time-consuming call finishes or not, `keyEvent` will be on another thread?

**Comment by wengxt** [comment, marked as answer]
no, fcitx is mostly single thread and uses main thread to process key event. it’s not like every key event will open a new thread

which means, if you block it, you will also block other app to fcitx event and everything. Please don’t do that.

**Reply by qingxiang-jia** [reply]
Hi @wengxt , sorry I asked this question because if I place `FCITX_INFO << keyEvent - begin;` at the beginning of the `keyEvent` function and `FCITX_INFO << keyEvent - end;` at the end of it, when I type, I would have output like the following:

```
keyEvent - begin
keyEvent - begin
keyEvent - begin
keyEvent - begin
some other thing
keyEvent - end
keyEvent - end
keyEvent - end
keyEvent - end
```

So I guessed it was running concurrently. I don't know how to explain this output if `keyEvent` is only called in one thread.

**Reply by wengxt** [reply]
show me you diff, you're likely not adding log in the right place.

**Reply by qingxiang-jia** [reply]
Will do that when I go home. Thanks!

**Reply by qingxiang-jia** [reply]
Hi @wengxt , here is the diff: https://github.com/qingxiang-jia/full-cloud-pinyin/compare/main...is-keyevent-async

**Reply by wengxt** [reply]
我觉得你应该注意到 keyEvent 有好多是提前返回而不会执行到 end 了吧。其实你想验证非常简单，只需要 --verbose=key_trace=5 来启动自然就可以看到所有按键事件的日志

**Reply by qingxiang-jia** [reply]
谢谢提醒, 之前没想到
